fe1-web/src/features/evoting/network/ElectionMessageApi.ts

```diff
@@ -41,46 +41,62 @@ export function requestCreateElection(
     questions: questions,
   });
 
-  const laoCh = channelFromIds(currentLao.id);
-  return publish(laoCh, message);
+  // publish on the general LAO channel
+  return publish(channelFromIds(currentLao.id), message);
 }
 
 /**
  * Sends a query to cast a vote during an election.
  *
- * @param election_id - The id of the ongoing election
- * @param votes - The votes to be added
+ * @param election - The election for which the vote should be casted
+ * @param votes - The votes to be added, a map from question index to the set of selected answer indices
  */
-export function castVote(election_id: Hash, votes: Vote[]): Promise<void> {
+export function castVote(
+  election: Election,
+  votes: { [questionIndex: number]: Set<number> },
+): Promise<void> {
   const time: Timestamp = Timestamp.EpochNow();
   const currentLao: Lao = OpenedLaoStore.get();
   const message = new CastVote({
     lao: currentLao.id,
-    election: election_id,
+    election: election.id,
     created_at: time,
-    votes: votes,
+    // Convert object to array
+    votes: Object.entries(votes)
+      .map(([idx, selectionOptions]) => ({
+        index: parseInt(idx, 10),
+        selectionOptions,
+      }))
+      // sort it by index
+      .sort((a, b) => a.index - b.index)
+      // and add an id to all votes as well as the matching question id
+      .map<Vote>(({ index, selectionOptions }) => ({
+        id: CastVote.computeVoteId(election, index, selectionOptions).valueOf(),
+        question: election.questions[index].id,
+        // sort votes in ascending order
+        vote: [...selectionOptions].sort((a, b) => a - b),
+      })),
   });
```

Comment:
This code looks very smart, but in general you want your code to look boring.
The more boring and obvious it is, the less effort it requires to maintain 😁 

Because the data processing logic is nested inside the object creation, it's both very hard to read and nearly impossible to step through in a debugger. I would strongly suggest to:
1. split this in multiple separate statements.
2. consider hiding the data manipulation in a dedicated function, taking in your `votes` variable and returning your `Vote[]` result.

Reply:
👍